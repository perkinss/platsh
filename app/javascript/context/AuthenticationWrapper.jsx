import React, { useEffect, useState, useContext } from 'react';
import User from '../model/user'

export const RootAuthContext = React.createContext();
export default ({ children }) => {
  let anyAuthVal,anyAuthData;
  try {
    anyAuthVal=JSON.parse(window.sessionStorage.getItem('authenticated'));
  }
  catch(err) {
    anyAuthVal=false;
  }
  try {
    anyAuthData = JSON.parse(window.sessionStorage.getItem('authData'))
  }
  catch(err) {
    anyAuthData = null;
  }
  const guestUser = new User({})
  const prevAuth = (anyAuthVal ? anyAuthVal : false);
  const prevAuthData = (anyAuthData ? anyAuthData : null);
  const prevUser = (anyAuthData ? new User(anyAuthData) : guestUser);
  const [authenticated, setAuthenticated] = useState(prevAuth);
  const [authData, setAuthData] = useState(prevAuthData);
  const [currentUser, setCurrentUser] = useState(prevUser);
  useEffect(
    () => {
      window.sessionStorage.setItem('authenticated', JSON.stringify(authenticated));
      window.sessionStorage.setItem('authData', (authData ? JSON.stringify(authData):null));
      setCurrentUser(authData ? new User(authData) : guestUser)
    },
    [authenticated, authData]
  );
  const defaultContext = {
    authenticated, setAuthenticated,
    authData, setAuthData,
    currentUser, setCurrentUser
  };
  return (
    <RootAuthContext.Provider value={defaultContext}>
      {children}
    </RootAuthContext.Provider>
  );
};
