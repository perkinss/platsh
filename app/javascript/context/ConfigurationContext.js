import React from 'react'

export const configuration = {
    sections: [
        {
            id: null,
            courses: [],
            name: '',
            assessment: [
                {
                    type: 'test',
                    name: '',
                    tasks: [
                    {
                        name: '',
                        available_standards: [
                            {
                                id: -1,
                                level: ''
                            }
                        ],
                        competencies: [

                        ]
                    },
                    ]
                }
            ],
            enrollees: [
                1, 2, 3, 22
            ]
        },
    ],

}


const ConfigurationContext =  React.createContext(configuration);

export { ConfigurationContext }
