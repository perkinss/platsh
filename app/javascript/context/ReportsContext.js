import React from 'react'

export const reports = {
    students: [],
    sections: [],
    selected_section: null,
    selected_student: null,
    section_overview_marks: {},
    section_competency_scores: {},
    student_section_marks: {},
    heat_map_colors: []
}


const ReportsContext =  React.createContext(reports);

export { ReportsContext }
