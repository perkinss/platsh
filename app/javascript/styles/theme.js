import {createMuiTheme} from "@material-ui/core/styles";

const theme = createMuiTheme({
    palette: {
        primary: {
            main: '#512da8',
        },
        secondary: {
            main: '#ab47bc',
        },
        type: 'dark'
    },
    slider: {
        selectionColor: '#F76B1C',
        handleFillColor: '#F76B1C',
        thumb: 'secondary'
    }

})

export default theme
