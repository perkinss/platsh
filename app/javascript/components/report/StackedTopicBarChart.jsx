import React, {useState} from 'react'
import ReactApexChart from 'react-apexcharts'
import {useTheme} from "@material-ui/styles/index";
import {CircularProgress} from "@material-ui/core";
import Typography from "@material-ui/core/Typography";

export default function StackedTopicBarChart(props) {
    const theme = useTheme()

    const { series, categories, calcStudentGrade, barHeight, loading } = props

    if (loading){
        return (
            <>
                <CircularProgress className={theme.progress} color={'secondary'} />
                <Typography variant={"body2"}>Topics Data Loading...</Typography>
            </>
        )
    }
    let chartHeight = barHeight
    if (categories && categories.length > 0) {
        chartHeight = categories.length * barHeight
    }
    const hide = window.innerWidth < 600
    const options = {
        chart: {
            stacked: true,
            foreColor: 'lightgrey',
            background: theme.palette.grey[800],
        },
        grid: {
          show: true,
          borderColor: theme.palette.grey[700]
        },
        plotOptions: {
            bar: {
                horizontal: true,
                barHeight: '90%',
                dataLabels: {
                    position: 'center',
                    enabledOnSeries: true,
                    maxItems: 100,
                    hideOverflowingLabels: false,
                }
            },
        },
        xaxis: {
            show: false,
            categories: categories,
            labels: {
                show: false
            },
            axisBorder: {
                show: false
            },
            axisTicks: {
                show: false
            }
        },
        yaxis: {
          labels: {
              formatter: (val) => {
                 if (typeof val === 'number') {
                     return val
                 }
                 return val.length < 27 ? val : val.substring(0,26) + '...'
              }
          }
        },
        dataLabels: {
            formatter: function (val, opts) {
                return val.toFixed(0) + "%"
            },
            textAnchor: 'middle',
        },
        tooltip: {
            theme: 'dark',
            y: {
                formatter: function (val) {
                    return val.toFixed(1) + " %"
                }
            },
            x: {
                show: true,
                formatter: (val, {series, seriesIndex, dataPointIndex, w}) => {
                    let grade = calcStudentGrade(val,dataPointIndex)
                    let msg = grade ? `${val}: Content grade: ${grade} %` : val
                    return msg
                }
            }
        },
        fill: {
            opacity: 1
        },
        legend: {
            showForSingleSeries: true,
            position: 'bottom',
            horizontalAlign: 'left',
            offsetX: 60, height: 100
        },
        theme: {
            mode: 'dark',
            palette: 'palette1',
        },
        responsive: [{
            breakpoint: 600,
            options: {
                chart: {
                    width: 350,
                    height:  chartHeight + 600
                },
                legend: {
                    show: false
                }
            }
        }]
    }

    return (
        <div id="stacked-topic-chart">
            <Typography variant={'h5'}>Content</Typography>
            { (series && series.length > 0) ?
                <ReactApexChart options={options} series={series} type="bar" height={`${chartHeight}`} />
                :
                <Typography variant={"caption"}>No data available</Typography>
            }
        </div>
    );
}