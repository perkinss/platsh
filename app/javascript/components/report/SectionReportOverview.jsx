import React, {useContext, useEffect, useState} from 'react'
import {Typography, Tabs, Tab, Grid, CircularProgress} from '@material-ui/core'
import SwipeableViews from 'react-swipeable-views';
import StackedTopicBarChart from "./StackedTopicBarChart";
import { useTheme } from "@material-ui/core/styles/index";
import StackedCompetencyBarChart from "./StackedCompetencyBarChart";
import {isNullOrUndefined} from "../../helpers/object_helper";
import {ReportsContext} from "../../context/ReportsContext";
import Slide from "@material-ui/core/Slide";
import Collapse from "@material-ui/core/Collapse";
import {getDateParams} from "../../helpers/date_helper";
import {fetchSectionCompetencyScores, fetchSectionTopicMarks} from "../../api/reportApi";

export default function SectionReportOverview (props) {
    const { classes, section, barHeight, hide, fromDate, toDate } = props
    const theme = useTheme();

    if (hide) {
        return ""
    }
    const { state, dispatch } = useContext(ReportsContext)

    const [tabIndex, setTabIndex] = useState(0);
    const [contentLabels, setContentLabels] = useState([])
    const [competencyLabels, setCompetencyLabels] = useState([])
    const [contentMarks, setContentMarks] = useState([])
    const [averages, setAverages] = useState([])
    const [competencyMarks, setCompetencyMarks] = useState([])
    const [competencyAverages, setCompetencyAverages] = useState([])

    const loadTopicsData = () => {
        let courseId = section.courses[tabIndex].id
        let course_data = state.section_overview_marks[section.id].section_content_overview.find( course => course.id === courseId )
        let marks = course_data.course_marks.filter(content => content.content_marks.length > 0)
        let contentSeries = marks.map ( content => {
            let data = content.content_marks.map( mark => mark.mark === null ? 0 : mark.mark.toFixed(2))
            let labels = content.content_marks.map( mark => mark.name)
            setContentLabels(labels)
            return { name: content.name, data: data }
        })
        setContentMarks(contentSeries)
        setAverages(course_data.averages)
    }

    const loadCompetencyData = () => {
        let courseId = section.courses[tabIndex].id
        let course_data = state.section_competency_scores[section.id].section_competency_overview.find( course => course.id === courseId )
        let marks = course_data.course_competencies.filter(group => group.competency_mark.length > 0)
        let compSeries = marks.map ( group => {
            let data = group.competency_mark.map( mark => mark.mark === null ? 0 : mark.mark.toFixed(1))
            let labels = group.competency_mark.map( mark => mark.name)
            setCompetencyLabels(labels)
            return { name: group.title, data: data }
        })
        setCompetencyMarks(compSeries)
        setCompetencyAverages(course_data.averages)
    }

    const shouldFetch = (slice) => {
        return section && !slice || !slice[section.id]
    }

    useEffect(()=> {
        if (!isNullOrUndefined(tabIndex)) {
            if (!shouldFetch(state.section_overview_marks)) {
                loadTopicsData()
            }
        }
    }, [tabIndex, state.section_overview_marks])

    useEffect(() => {
        if (!isNullOrUndefined(tabIndex)) {
            if (!shouldFetch(state.section_competency_scores)) {
                loadCompetencyData()
            }
        }
    }, [tabIndex, state.section_competency_scores])

    useEffect(() => {
        if (section) {
            if (shouldFetch(state.section_overview_marks)) {
                fetchSectionTopicMarks(dispatch, state, fromDate, toDate)
            } else {
                loadTopicsData()
            }
            if (shouldFetch(state.section_competency_scores)) {
                fetchSectionCompetencyScores(dispatch, state, fromDate, toDate)
            } else {
                loadCompetencyData()
            }
        }
    }, [section])

    const handleChangeIndex = (index) => {
        setTabIndex(index)
    }

    const handleTabChange = (event, newValue) => {
        setTabIndex(newValue)
    }

    const calcContentGrade = (val, dataPointIndex) => {
        if ( averages[dataPointIndex]) {
            return averages[dataPointIndex].toFixed(1)
        }
    }

    const calcCompetencyGrade = (val, dataPointIndex) => {
        if (competencyAverages[dataPointIndex]) {
            return competencyAverages[dataPointIndex].toFixed(0)
        }
    }

    return (
        <>
            <Tabs
                value={tabIndex}
                onChange={handleTabChange}
                indicatorColor="secondary"
                textColor="secondary"
                variant="fullWidth"
            >
                {section.courses.map(course => {
                    return <Tab label={course.title} key={`tab-${course.title}`}/>
                })}
            </Tabs>
            <SwipeableViews
                axis={theme.direction === 'rtl' ? 'x-reverse' : 'x'}
                index={tabIndex}
                onChangeIndex={handleChangeIndex}
                style={{ overflow: 'hidden', padding: 20 }}
            >
                {section.courses.map(course => {
                    return <TabContainer dir={theme.direction} key={`view-${course.title}`} className={classes.content}>
                        <Grid container spacing={0}>
                            <Grid item xs={12} sm={8} style={{padding: '5px 10px'}}>
                                <StackedTopicBarChart
                                    series={contentMarks}
                                    categories={contentLabels}
                                    calcStudentGrade={calcContentGrade}
                                    barHeight={barHeight}
                                    loading={state.section_loading.topics}
                                />
                            </Grid>
                            <Grid item xs={12} sm={4} style={{padding: '5px 10px' }}>
                                <StackedCompetencyBarChart
                                    series={competencyMarks}
                                    categories={competencyLabels}
                                    calcStudentGrade={calcCompetencyGrade}
                                    barHeight={barHeight}
                                    loading={state.section_loading.competency}
                                />
                            </Grid>
                        </Grid>
                    </TabContainer>
                })}
            </SwipeableViews>
        </>
    )
}

function TabContainer({ children, dir }) {
    return (
        <Typography component="div" dir={dir} >
            {children}
        </Typography>
    );
}