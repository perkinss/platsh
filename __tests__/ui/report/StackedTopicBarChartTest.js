import React, { useState } from 'react'
import { createRender, createShallow } from '@material-ui/core/test-utils';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import expect from 'expect'
import { ConfigurationContext } from "../../../app/javascript/context/ConfigurationContext";
import theme from "../../../app/javascript/styles/theme";
import { ThemeProvider } from '@material-ui/styles'
import TeachersStudentReport from "../../../app/javascript/components/report/TeachersStudentReport";
import SectionReportOverview from "../../../app/javascript/components/report/SectionReportOverview";
import StackedTopicBarChart from "../../../app/javascript/components/report/StackedTopicBarChart";

Enzyme.configure({ adapter: new Adapter() });

describe('<StackedTopicBarChart />', () => {
    let render;
    let classes = {
        content: {}
    }
    let state = {
    }
    const series =   [{
        name: 'Operations',
        data: [12.5, 10, 12.5, 8, 7, 12.5, 12.5]
    }, {
        name: 'Exponents',
        data: [10, 12.5, 12.5, 12.5, 12.5, 12.5, 12.5]
    }, {
        name: 'Polynomials',
        data: [12.5, 12.5, 11, 9, 12.5, 12.5, 12.5]
    }, {
        name: 'two-variable linear relations',
        data: [12.5, 7, 12.5, 8, 6, 12.5, 12.5]
    }, {
        name: 'linear equations',
        data: [12.5, 12, 12.5, 12.5, 12, 12.5, 12]
    },
        {
            name: 'proportional reasoning',
            data: [12.5, 5, 5, 10, 10, 8, 5]
        },
        {
            name: 'statistics',
            data: [10, 2, 0, 0, 3, 0, 5]
        },
        {
            name: 'financial literacy',
            data: [12, 1, 1, 3,1, 3, 3]
        }
    ]

    const categories = ['Liz', 'Ashlee Yin', 'Michael Dorn', 'Worf', 'Buzz', 'Liz', 'Zoe']

    let section = {
        name: 'New at the Zoo...',
        courses: [
            {
                title: 'Math 9'
            }
        ]
    }
    beforeAll(() => {
        render = createRender();
    })

    it('should match the snapshot', () => {
        const wrapper = render(
            <ThemeProvider theme={theme}>
                <ConfigurationContext.Provider value={{dispatch: () => {}, state: state }} >
                    <StackedTopicBarChart
                        classes={classes}
                        section={section}
                        series={series}
                        categories={categories}
                    />
                </ConfigurationContext.Provider>
            </ThemeProvider>
                );
        expect(wrapper).not.toBeNull()
        expect(wrapper).toMatchSnapshot()
    });

    it('should match the snapshot', () => {
        const wrapper = render(
            <ThemeProvider theme={theme}>
                <ConfigurationContext.Provider value={{dispatch: () => {}, state: state }} >
                    <StackedTopicBarChart
                        classes={classes}
                        section={section}
                        series={series}
                        categories={categories}
                    />
                </ConfigurationContext.Provider>
            </ThemeProvider>
        );
        let chart = wrapper.find('#chart')
        expect(chart).not.toBeNull()
    });

    it('should be empty if there is no data', () => {
        const wrapper = render(
            <ThemeProvider theme={theme}>
                <ConfigurationContext.Provider value={{dispatch: () => {}, state: state }} >
                    <StackedTopicBarChart
                        classes={classes}
                        section={section}
                        series={[]}
                        categories={categories}
                    />
                </ConfigurationContext.Provider>
            </ThemeProvider>
        );
        let chart = wrapper.find('#chart')
        expect(chart).not.toBeNull()
        expect(wrapper.html()).toContain('No data available')

    });

    it('should have a spinner when loading', () => {
        const wrapper = render(
            <ThemeProvider theme={theme}>
                <ConfigurationContext.Provider value={{dispatch: () => {}, state: state }} >
                    <StackedTopicBarChart
                        classes={classes}
                        section={section}
                        series={[]}
                        categories={categories}
                        loading={true}
                    />
                </ConfigurationContext.Provider>
            </ThemeProvider>
        );
        let chart = wrapper.find('#chart')
        expect(chart).not.toBeNull()
        expect(wrapper.html()).toMatch(/CircularProgress/)

    });

})