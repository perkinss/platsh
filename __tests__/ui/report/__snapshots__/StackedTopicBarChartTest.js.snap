// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StackedTopicBarChart /> should match the snapshot 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "id": "stacked-topic-chart",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "class": "MuiTypography-root MuiTypography-h5",
        },
        "children": Array [
          Node {
            "data": "Content",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "h5",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {
            "height": "NaN",
            "options": "[object Object]",
            "series": "[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]",
            "type": "bar",
            "width": "100%",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "height": undefined,
            "options": undefined,
            "series": undefined,
            "type": undefined,
            "width": undefined,
          },
          "x-attribsPrefix": Object {
            "height": undefined,
            "options": undefined,
            "series": undefined,
            "type": undefined,
            "width": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Node {
        "attribs": Object {
          "height": "NaN",
          "options": "[object Object]",
          "series": "[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]",
          "type": "bar",
          "width": "100%",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {
            "class": "MuiTypography-root MuiTypography-h5",
          },
          "children": Array [
            Node {
              "data": "Content",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "h5",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "height": undefined,
          "options": undefined,
          "series": undefined,
          "type": undefined,
          "width": undefined,
        },
        "x-attribsPrefix": Object {
          "height": undefined,
          "options": undefined,
          "series": undefined,
          "type": undefined,
          "width": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "id": undefined,
    },
    "x-attribsPrefix": Object {
      "id": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
